merqube_client_lib/templates/equity_baskets/util.py

```diff
@@ -0,0 +1,194 @@
+)
+
+SPEC_KEYS = ["base_date"]
+TOP_LEVEL = ["namespace", "name", "title", "base_date"]
+MISC = ["apikey", "run_hour", "run_minute"]
+INFO_REQUIRED = TOP_LEVEL + SPEC_KEYS + MISC
+OPTIONAL = "timezone"
+
+
+logger = get_module_logger(__name__, level=logging.DEBUG)
+
+
+def inline_to_tp(index_model: IndexDefinitionPatchPutGet) -> list[tuple[str, EquityBasketPortfolio]]:
+    """Convert the inline portfolio to a list of target portfolios"""
+
+    target_portfolios = []
+
+    assert index_model.spec and index_model.spec.index_class_args and index_model.spec.index_class_args.get("spec")
+    portfolio = index_model.spec.index_class_args["spec"]["portfolios"]
+    uom = PortfolioUom.UNITS if portfolio["quantity_type"] == "SHARES" else PortfolioUom.WEIGHT
+    id_type = portfolio["identifier_type"]
+
+    # see how nany TP values we need
+    dates = set()
+    for constituent in (const := portfolio["constituents"]):
+        dates.add(constituent["date"])
+
+    for date in sorted(dates):
+        positions: list[BasketPosition | RicEquityPosition] = []
+        for constituent in const:
+            sec_type = constituent["security_type"]
+            pos_class = RicEquityPosition if sec_type == "EQUITY" else BasketPosition
+
+            if constituent["date"] != date:
+                continue  # will get picked up in another target portfolio
+
+            positions.append(
+                pos_class(
+                    asset_type=constituent["security_type"],
+                    identifier=constituent["identifier"],
+                    amount=constituent["quantity"],
+                    # this is only at the top level of inline; but its wrong for the cash position (in TP it is CURRENCY_CODE)
+                    identifier_type="CURRENCY_CODE" if sec_type == "CASH" else id_type,
+                )
+            )
+
+        target_port_val = EquityBasketPortfolio(timestamp=date, unit_of_measure=uom, positions=positions)
+
+        EquityBasketPortfolio.parse_obj(target_port_val)  # validate
+        target_portfolios.append((date, target_port_val))
+
+    return target_portfolios
+
+
+def read_file(filename: str) -> list[Any]:
+    """reads portfolio/overrides files"""
+    return list(pd.read_csv(filename, float_precision="round_trip").to_dict(orient="index").values())
+
+
+def log_index(index: IndexDefinitionPost) -> None:
+    """
+    Logs the resulting index
+    """
+    json_formatted_str = json.dumps(json.loads(index.json(exclude_none=True)), indent=4)
+    logger.info("Index spec: \n" + json_formatted_str)
+
+
+def get_index_info(config_file_path: str, required_fields: list[str]) -> dict[str, Any]:
+    """
+    load index specific info from config file + validate the data
+    """
+    index_info = cast(dict[str, Any], json.load(open(config_file_path, "r")))
+
+    for k in required_fields:
+        if k not in index_info:
+            raise ValueError(f"Missing key {k} in index info")
+
+    try:
+        for k in ["base_date"]:
```

Comment:
Why loop?

Reply:
was originally a longer list - ill make this a DATE_KEYS variable